src/fsm.c

```diff
@@ -395,37 +479,115 @@ static int decodeDatabase(struct fsm *f, struct cursor *cursor)
 	return 0;
 }
 
+static unsigned dbNumPages(struct db *db)
+{
+	sqlite3_vfs *vfs;
+	int rv;
+	uint32_t n;
+
+	vfs = sqlite3_vfs_find(db->config->name);
+	rv = VfsDatabaseNumPages(vfs, db->filename, &n);
+	assert(rv == 0);
+	return n;
+}
+
+/* Determine the total number of raft buffers needed for a snapshot */
+static unsigned snapshotNumBufs(struct fsm *f)
+{
+	struct db *db;
+	queue *head;
+	unsigned n = 1; /* snapshot header */
+
+	QUEUE__FOREACH(head, &f->registry->dbs)
+	{
+		n += 2; /* database header & wal */
+		db = QUEUE__DATA(head, struct db, queue);
+		n += dbNumPages(db); /* 1 buffer per page (zero copy) */
+	}
+
+	return n;
+}
+
+/* An example array of snapshot buffers looks like this:
+ *
+ * bufs:  SH DH1 P1 P2 P3 WAL1 DH2 P1 P2 WAL2
+ * index:  0   1  2  3  4    5   6  7  8    9
+ *
+ * SH:   Snapshot Header
+ * DHx:  Database Header
+ * Px:   Database Page (not to be freed)
+ * WALx: a WAL
+ * */
+static void freeSnapshotBufs(struct fsm *f,
+			     struct raft_buffer bufs[],
+			     unsigned n_bufs)
+{
+	queue *head;
+	struct db *db;
+	unsigned i;
+
+	if (bufs == NULL || n_bufs == 0) {
+		return;
+	}
+
+	/* Free snapshot header */
+	sqlite3_free(bufs[0].base);
+
+	i = 1;
+	/* Free all database headers & WAL buffers */
+	QUEUE__FOREACH(head, &f->registry->dbs)
+	{
+		if (i == n_bufs) {
+		    break;
+		}
+		db = QUEUE__DATA(head, struct db, queue);
+		/* i is the index of the database header */
+		sqlite3_free(bufs[i].base);
+		/* i is now the index of the next database header (if any) */
+		i += 1 + dbNumPages(db) + 1;
```

Comment:
Would be nice to add a comment about what those `1`s are (I guess the two bueffers for the database header and WAL). E.g.

```C
i += 1 /* db header */ + dbNumPages(db) + 1 /* WAL */;
```